scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -565,6 +590,13 @@ class ScioContext private[scio] (val options: PipelineOptions, private var artif
 
     val closedContext = ClosedScioContext(this.pipeline.run(), this)
 
+    _onJobStart(closedContext)
+
+    if (_hasJobFinishHook) {
+      val result = closedContext.waitUntilFinish(closedContext.getAwaitDuration)
```

Comment:
But won't this now make `close()` turn into a blocking call simply by using `onJobFinish` once? I think we need to look deeper for a more appropriate place to trigger the `_onJobFinish` call. We might have to make a choice between:
1. Only trigger the `_onJobFinish` call if the user actually blocks on the job to finish, either by calling `close()` with `isBlocking == true` or explicitly calling `waitUntilDone` or `waitUntilFinish` on the ClosedScioContext. If they don't do anything like that, the event never triggers, even if they `sleep` for a year after calling (non-blocking) `close()`.
2. Use a background thread to poll job progress.

I'm don't know much about the code base, so I can't say what's the best way. There could be other ways to do it. But I think turining `close()` into a blocking call simply because there's an even handler registered will be surprising to users; it's too implicit, even if documented.

Reply:
true that, but isn't a registered callback never being called unless they explicitly wait also kind of surprising?

Reply:
then again, i think we technically don't really need this, it is possible to have a `.close().waitUntilDone().doWhatever()` in the pipeline. Closing this PR.